File: alljoyn_c/src/BusListener.cc
/**
 * @file
 * This implements the C accessable version of the BusListener class using
 * function pointers, and a pass-through implementation of BusListener.
 */

/******************************************************************************
 * Copyright (c) 2009-2014 AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <alljoyn/BusListener.h>
#include <alljoyn/MsgArg.h>
#include <alljoyn_c/BusListener.h>
#include <string.h>
#include <assert.h>
#include "DeferredCallback.h"
#include <qcc/Debug.h>

#define QCC_MODULE "ALLJOYN_C"

static void AJ_CALL __PropertyChanged(alljoyn_buslistener_bus_prop_changed_ptr fcn, const void* context, const char* prop_name, alljoyn_msgarg prop_value)
{
    fcn(context, prop_name, prop_value);

    if (prop_value) {
        alljoyn_msgarg_destroy(prop_value);
    }
}


namespace ajn {

/**
 * Abstract base class implemented by AllJoyn users and called by AllJoyn to inform
 * users of bus related events.
 */
class BusListenerCallbackC : BusListener {
  public:
    BusListenerCallbackC(const alljoyn_buslistener_callbacks* in_callbacks, const void* in_context)
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        memcpy(&callbacks, in_callbacks, sizeof(alljoyn_buslistener_callbacks));
        context = in_context;
    }

    void ListenerRegistered(BusAttachment* bus)
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.listener_registered != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.listener_registered(context, (alljoyn_busattachment)bus);
            } else {
                DeferredCallback_2<void, const void*, alljoyn_busattachment>* dcb =
                    new DeferredCallback_2<void, const void*, alljoyn_busattachment>(callbacks.listener_registered, context, (alljoyn_busattachment)bus);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void ListenerUnregistered()
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.listener_unregistered != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.listener_unregistered(context);
            } else {
                DeferredCallback_1<void, const void*>* dcb =
                    new DeferredCallback_1<void, const void*>(callbacks.listener_unregistered, context);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void FoundAdvertisedName(const char* name, TransportMask transport, const char* namePrefix)
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.found_advertised_name != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.found_advertised_name(context, name, transport, namePrefix);
            } else {
                DeferredCallback_4<void, const void*, const char*, TransportMask, const char*>* dcb =
                    new DeferredCallback_4<void, const void*, const char*, TransportMask, const char*>(callbacks.found_advertised_name, context, name, transport, namePrefix);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void LostAdvertisedName(const char* name, TransportMask transport, const char* namePrefix)
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.lost_advertised_name != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.lost_advertised_name(context, name, transport, namePrefix);
            } else {
                DeferredCallback_4<void, const void*, const char*, TransportMask, const char*>* dcb =
                    new DeferredCallback_4<void, const void*, const char*, TransportMask, const char*>(callbacks.lost_advertised_name, context, name, transport, namePrefix);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void NameOwnerChanged(const char* busName, const char* previousOwner, const char* newOwner)
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.name_owner_changed != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.name_owner_changed(context, busName, previousOwner, newOwner);
            } else {
                DeferredCallback_4<void, const void*, const char*, const char*, const char*>* dcb =
                    new DeferredCallback_4<void, const void*, const char*, const char*, const char*>(callbacks.name_owner_changed, context, busName, previousOwner, newOwner);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void BusStopping()
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.bus_stopping != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.bus_stopping(context);
            } else {
                DeferredCallback_1<void, const void*>* dcb =
                    new DeferredCallback_1<void, const void*>(callbacks.bus_stopping, context);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void BusDisconnected()
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.bus_disconnected != NULL) {
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                callbacks.bus_disconnected(context);
            } else {
                DeferredCallback_1<void, const void*>* dcb =
                    new DeferredCallback_1<void, const void*>(callbacks.bus_disconnected, context);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

    void PropertyChanged(const char* propName, const MsgArg* propValue)
    {
        QCC_DbgTrace(("%s", __FUNCTION__));
        if (callbacks.property_changed != NULL) {
            alljoyn_msgarg msg_arg = propValue ? alljoyn_msgarg_create_and_set(propValue->Signature().c_str(), propValue->v_variant.val) : NULL;
            // must wrap the user function in order to properly clean up msg_arg
            if (!DeferredCallback::sMainThreadCallbacksOnly) {
                __PropertyChanged(callbacks.property_changed, context, propName, msg_arg);
            } else {
                DeferredCallback_4<void, alljoyn_buslistener_bus_prop_changed_ptr, const void*, const char*, alljoyn_msgarg>* dcb =
                    new DeferredCallback_4<void, alljoyn_buslistener_bus_prop_changed_ptr, const void*, const char*, alljoyn_msgarg>(
                        __PropertyChanged, callbacks.property_changed, context, propName, msg_arg);
                DEFERRED_CALLBACK_EXECUTE(dcb);
            }
        }
    }

  protected:
    alljoyn_buslistener_callbacks callbacks;
    const void* context;
};

}

struct _alljoyn_buslistener_handle {
    /* Empty by design, this is just to allow the type restrictions to save coders from themselves */
};

alljoyn_buslistener AJ_CALL alljoyn_buslistener_create(const alljoyn_buslistener_callbacks* callbacks, const void* context)
{
    QCC_DbgTrace(("%s", __FUNCTION__));
    return (alljoyn_buslistener) new ajn::BusListenerCallbackC(callbacks, context);
}

void AJ_CALL alljoyn_buslistener_destroy(alljoyn_buslistener listener)
{
    QCC_DbgTrace(("%s", __FUNCTION__));
    assert(listener != NULL && "listener parameter must not be NULL");
    delete (ajn::BusListenerCallbackC*)listener;
}

File: alljoyn_java/src/org/alljoyn/bus/BusListener.java
/*
 * Copyright (c) 2009-2011, 2014 AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package org.alljoyn.bus;

/**
 * Bus listeners are responsible for handling callbacks from the AllJoyn system.
 * It is expected that a user of the AllJoyn bus will specialize this class in
 * order to handle required events from the bus.
 *
 * Listener objects are the Java objects that handle notification events and are
 * called from AllJoyn in the context of one of its threads.  All listener
 * objects are expected to be multithread safe (MT-Safe) between construction
 * and destruction.  That is, every thread executing in a listener object's
 * methods 1) gets a unique copy of all temporary data (it is re-entrant); and
 * 2) all shared data -- the object instance's member variables or any globals
 * must contain no read-modify-write access patterns (okay to write or read,
 * just never to read-modify-write).  If such access patterns are required, it
 * is the responsibility of the client to, for example, add the synchronized
 * keyword when overriding one of the listener methods or provide some other
 * serialization mechanism in order to preserve MT-Safe operation.
 *
 * This rule extends to other objects accessed during processing of
 * notifications.  For example, it is a programming error to allow a notifiation
 * method to update a collection in another object without serializing access
 * to the collection.
 *
 * The important consideration in this case is that as soon as one sets up a
 * listener object to receive notifications from AllJoyn, one is implicitly
 * dealing with multithreaded code.
 *
 * Since listener objects generally run in the context of the AllJoyn thread
 * which manages reception of events, If a blocking AllJoyn call is made in
 * the context of a notification, the necessary and sufficient conditions for
 * deadlock are established.
 *
 * The important consideration in this case is that when one receives a
 * notification from AllJoyn, that notification is executing in the context of
 * an AllJoyn thread.  If one makes a blocking call back into AllJoyn on that
 * thread, a deadlock cycle is likely, and if this happens your bus attachment
 * receive thread will deadlock (with itself).  The deadlock is typically broken
 * after a bus timeout eventually happens.
 */
public class BusListener {

    /**
     * Destroy native resources held by objects of this class.
     */
    public BusListener() {
        create();
    }

    /**
     * Destroy native resources held by objects of this class.
     */
    protected void finalize() throws Throwable {
        try {
            destroy();
        } finally {
            super.finalize();
        }
    }

    /**
     * Create any native resources held by objects of this class.  Specifically,
     * we allocate a C++ counterpart of this listener object.
     */
    private native void create();

    /**
     * Release any native resources held by objects of this class.
     * Specifically, we may delete a C++ counterpart of this listener object.
     */
    private native void destroy();

    /**
     * Called by the bus when the listener is registered. This give the listener implementation the
     * opportunity to save a reference to the bus.
     *
     * @param bus  The bus the listener is registered with.
     */
    public void listenerRegistered(BusAttachment bus) {}

    /**
     * Called by the bus when the listener is unregistered.
     */
    public void listenerUnregistered() {}

    /**
     * Called by the bus when an external bus is discovered that is advertising
     * a well-known name that this attachment has registered interest in via a
     * call to findAdvertisedName
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     *
     * @param name         A well known name that the remote bus is advertising.
     * @param transport    Transport that received the advertisment.
     * @param namePrefix   The well-known name prefix used in call to
     *                     FindAdvertisedName that triggered this callback.
     */
    public void foundAdvertisedName(String name, short transport, String namePrefix) {}

    /**
     * Called by the bus when an advertisement previously reported through
     * foundAdvertisedName has become unavailable.
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     *
     * @param name         A well known name that the remote bus is advertising
     *                     that is of interest to this attachment.
     * @param transport    Transport that stopped receiving the given advertised name.
     * @param namePrefix   The well-known name prefix that was used in a call to
     *                     FindAdvertisedName that triggered this callback.
     */
    public void lostAdvertisedName(String name, short transport, String namePrefix) {}

    /**
     * Called by the bus when the ownership of any well-known name changes.
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     *
     * @param busName        The well-known name that has changed.
     * @param previousOwner  The unique name that previously owned the name or
     *                       NULL if there was no previous owner.
     * @param newOwner       The unique name that now owns the name or NULL if
     *                       there is no new owner.
     */
    public void nameOwnerChanged(String busName, String previousOwner, String newOwner) {}

    /**
     * @deprecated This is deprecated.  It is no longer called when properties change.  Use the property change notification mechanism for proxy bus objects instead.
     *
     * Called by the bus when the value of a property changes if that property has annotation
     * org.freedesktop.DBus.Properties.PropertiesChanged=true
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     *
     * @param propName       The well-known name that has changed.
     * @param propValue      The new value of the property; NULL if not present
     */
    @Deprecated
    public void propertyChanged(String propName, Variant propValue) {}

    /**
     * Called when a bus that this listener is registered with is stopping.
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     */
    public void busStopping() {}

    /**
     * Called when a BusAttachment this listener is registered with is has become disconnected from
     * the bus
     *
     * Any implementation of this function must be multithread safe.  See the
     * class documentation for details.
     */
    public void busDisconnected() {}


    /**
     * The opaque pointer to the underlying C++ object which is actually tied
     * to the AllJoyn code.
     */
    private long handle = 0;

}
